### PR TITLE
Add a native QR code generator

### PR DESCRIPTION
**Pull request by v16Studios**

### CODE REVIEW

**Review by leonardocustodio — COMMENTED** [review]
Are all qr codes generated for single use? I was wondering if it would make sense to delete the QR code after they are scanned to not fill all disk storage saving those (obviously it will take quite a while to fill the storage only with this kind of image)

**Comment by v16Studios** [comment]
> Are all qr codes generated for single use? I was wondering if it would make sense to delete the QR code after they are scanned to not fill all disk storage saving those (obviously it will take quite a while to fill the storage only with this kind of image)

The QRs aren't stored on the server at all, they're generated on-the-fly and sent directly to the client (or should be! 😅)

**Comment by leonardocustodio** [comment]
Oh, I saw the route and thought It was saving it and the route was to get/show it, I believe I misunderstood the code

**Review by leonardocustodio — APPROVED** [review]
I would only make sure with Brad if we have imagick installed in our hosted platform. I know NFTio did not have it in their env. Just in case people try to generate png on it

**Comment by v16Studios** [comment]
It's set to SVG by default in the config which I believe doesn't need Imagick, user's who implement this package (including ourselves) will decide if they want to generate PNGs and can install Imagick on their servers 🙂👍🏻

**Comment by leonardocustodio** [comment]
Yep, just saying we should probably get Brad to install it if it is not installed as people usually prefer png over svg, even though svg is way better

**Review by leonardocustodio — APPROVED** [review]

File: src/Support/Util.php [hunk 1]
```diff
@@ -16,4 +16,13 @@ public static function createJsonRpc(string $method, array $params = []): string
             'id' => random_int(1, 999999999),
         ], JSON_THROW_ON_ERROR);
     }
+
+    public static function isBase64String(string $string)
+    {
+        if (base64_encode(base64_decode($string, true)) === $string) {
```
Comment (enjinabner) [inline src/Support/Util.php:22]:
```suggestion
 return base64_encode(base64_decode($string, true)) === $string
```

File: src/Http/Controllers/QrController.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        $qrCode = QRCode::format($format)->size($size)->generate($data);
+        $qrCode = base64_encode($qrCode);
+
+        $src = match ($format) {
+            'eps' => "data:application/postscript;base64,{$qrCode}",
+            'png' => "data:image/png;base64,{$qrCode}",
+            default => "data:image/svg+xml;base64,{$qrCode}",
+        };
+
+        return response("<img src='{$src}' alt='QR Code'>");
```
Comment (enjinabner) [inline src/Http/Controllers/QrController.php:39]: Should we add caching for performance?

File: src/Http/Controllers/QrController.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+class QrController extends Controller
+{
+    public function get(Request $request)
+    {
+        $data = $request->input('d');
+        $size = $request->input('s') ?? config('enjin-platform.qr.size');
+        $format = $request->input('f') ?? config('enjin-platform.qr.format');
+
+        if (empty($data)) {
+            throw new PlatformException(__('enjin-platform::error.qr.data_must_not_be_empty'), 501);
```
Comment (enjinabner) [inline src/Http/Controllers/QrController.php:19]: I don't see this added in the translation file

File: src/Http/Controllers/QrController.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        }
+
+        if (!in_array($format, ['eps', 'png', 'svg'])) {
+            throw new PlatformException(__('enjin-platform::error.qr.image_format_not_supported'), 501);
```
Comment (enjinabner) [inline src/Http/Controllers/QrController.php:23]: I don't see this added in the translation file

File: src/Http/Controllers/QrController.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        }
+
+        if ('png' == $format && !extension_loaded('Imagick')) {
+            throw new PlatformException(__('enjin-platform::error.qr.extension_not_installed'), 501);
```
Comment (enjinabner) [inline src/Http/Controllers/QrController.php:27]: I don't see this added in the translation file

**Review by enjinabner — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/QrController.php (hunk 2 above)
Comment (v16Studios) [reply src/Http/Controllers/QrController.php:39]: It's already a pretty lightweight function, we probably wouldn't see much benefit from caching.

**Review by enjinabner — APPROVED** [review]

**Review by leonardocustodio — APPROVED** [review]

**Pull request merged**